### DISCUSSION TITLE
Variable as parameter of custom function

### DISCUSSION THREAD

**Discussion by NullDev** [josdejong/mathjs#3156] [Q&A] [answered]
So, I'm building a simple sigmaSum function that I'm importing via `mathjs.import`

and the function works perfectly fine. e.g.: 

`sigmaSum("n=1", "k=5", "n^2 + k")`
`-> 80`

Now the problem is, because it's a custom function and I have to use quotes for the arguments, I can't actually pass variables. e.g.: this does not work: 

`f(n) = sigmaSum("i=1","k=n-1", "floor(1 / gcd(i, k))"); f(15)`

because `n` defined in `f(n)`, which thus should be 15, is unknown to `sigmaSum`. 
Is there a way to actually get the context and all variables defined "outside" of sigmaSum?
it would be ideal if I wouldn't need the quotes all together and can just parse the math as-is but I think that's not possible. 

I could live with `f(n) = sigmaSum("i=1", "k=${n}-1", "i + k"); f(10)` as well, but this doesn't seem to work. I'm not sure when and how to use toTex() with my code since its literally just 

```js
const sigmaSum = function(n, k, expr){
    const [n_Symbol, n_Value] = parseVar(n);
    const [k_Symbol, k_Value] = parseVar(k);   

    let result = 0;
    for (let i = n_Value; i <= k_Value; i++){
        result += mathjs.evaluate(expr.replaceAll(n_Symbol, i).replaceAll(k_Symbol, k_Value))
     }
     return result;
};

mathjs.import({
    sigmaSum,
});
```

any help would be appreciated :)

**Comment by dvd101x** [comment]
Hi, here are a few pointers

To access variables outside `sigmaSum` you need to use either a parser or a scope. I guess your `mathjs` isn't a parser. https://mathjs.org/docs/expressions/parsing.html

To be able to use `"k=${n}-1"` it seems you intend to use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) so you need to use back `` ` `` instead of `"` like so: `` `k=${n}-1` ``

Your function definition might be more usable if you don't assign it to a constant:

Instead of
```js
const sigmaSum = function(n, k, expr){
}
```
Try 
```js
function sigmaSum(n, k, expr){
  // this way you could call the function in lines before its definition
}
```
You could remove the for loop by using `math.sum` directly over an array.

Maybe use a helper function that always receives the arguments in the exact format.
```js
function summation(i0, n, f) {
  const I = math.range(i0, n, true)
  return math.sum(I.map(i => f(i, n)))
}
console.log(summation(1, 5, (n, k) => n ** 2 + k))                // 80
console.log(summation(1, 5, math.evaluate("f(n, k) = n^2 + k")))  // 80
```

**Reply by NullDev** [reply]
Hey! Thank you for your reply :) 

---

> I guess your mathjs isn't a parser.

That's correct. I'm using the evaluator for that (or rather, a [wrapper for `mathjs.evaluate()`](https://github.com/NullDev/Arithmetica-Bot/blob/master/src/util/mathEval.js#L236))

---

> To be able to use "k=${n}-1" it seems you intend to use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) so you need to use back ` instead of " like so: `k=${n}-1`

Template literals don't really work in that case. e.g. ``f(n) = sigmaSum("i=1",`k=${n}-1`, "floor(1 / gcd(i,k))"); f(15)`` doesn't actually interpolate the `n` from `f()` into my function. I was referring to `.toTex()` here: https://mathjs.org/docs/expressions/customization.html#custom-latex-handlers - which I thought could be an option. 

---

> this way you could call the function in lines before its definition

That's an intentional design choice. Hoisting can be rather counterproductive at times which is why I explicitly use const lambda functions

---

> You could remove the for loop by using math.sum directly over an array.

That was intentional as well because I limit the computation time of my iterative calculator function. I omitted that part in the code because it wasn't really relevant to my question. 

```js
const sigmaSum = function(n, k, expr){
    const [n_Symbol, n_Value] = parseVar(n);
    const [k_Symbol, k_Value] = parseVar(k);   

    let result = 0;
    const startTime = Date.now();

    for (let i = n_Value; i <= k_Value; i++){
        if (Date.now() - startTime > computationLimitSecs * 1000){
            throw new Error("Function execution exceeded " + computationLimitSecs + " seconds");
        }

        result += mathjs.evaluate(expr.replaceAll(n_Symbol, i).replaceAll(k_Symbol, k_Value))
     }
     return result;
};
```

a rather naive approach, but it's good enough I guess.
The whole thing is [here](https://github.com/NullDev/Arithmetica-Bot/blob/master/src/util/mathEval.js#L130-L167)

---

I really appreciate all the feedback though, thanks! 
Im just still not sure how I'd go about passing "outside" variables to my custom function.

**Reply by dvd101x** [reply]
Oh thanks for reviewing the reply and for sharing your project it's very cool.

I misunderstood the purpose of your code. Thanks for your clarifications.

**Comment by josdejong** [comment, marked as answer]
@NullDev you can use the `rawArgs` feature, then you can use `sigmaSum(n=1, k=5, n^2 + k)` and 
`f(n) = sigmaSum(i=1, k=n-1, floor(1 / gcd(i, k))); f(15)` just like you want.

Docs: https://mathjs.org/docs/expressions/customization.html#custom-argument-parsing

Here a full example: 

```js
import { create, all } from 'mathjs'

const mathjs = create(all)

function sigmaSum (args, math, scope){
  const [nNode, kNode, exprNode] = args

  if (!nNode.isAssignmentNode || !nNode.object.isSymbolNode) {
    throw Error('First argument must define a variable, like "i=1"')
  }
  const nName = nNode.object.name

  if (!kNode.isAssignmentNode || !kNode.object.isSymbolNode) {
    throw Error('Second argument must define a variable, like "k=5"')
  }
  const kName = kNode.object.name

  const n = nNode.compile().evaluate(scope);
  const k = kNode.compile().evaluate(scope);
  const expr = exprNode.compile()

  let result = 0;
  for (let i = n; i <= k; i++) {
    const newScope = new Map()
    scope.forEach((value, key) => newScope.set(key, value))
    newScope.set(nName, i)
    newScope.set(kName, k)
    result += expr.evaluate(newScope)
  }
  return result;
}
sigmaSum.rawArgs = true

mathjs.import({
  sigmaSum,
});

// Usage
const result1 = mathjs.evaluate('sigmaSum(n=1, k=5, n^2 + k)')
console.log('result1:', result1) // 80

const scope2 = new Map()
mathjs.evaluate('f(n) = sigmaSum(i=1, k=n-1, floor(1 / gcd(i, k)))', scope2)
const result2 = mathjs.evaluate('f(15)', scope2)
console.log('result2:', result2) // 6

const scope3 = new Map()
mathjs.evaluate('f(x) = sigmaSum(n=1, k=5, sigmaSum(a=1, b=5, x + a))', scope3)
const result3 = mathjs.evaluate('f(10)', scope3)
console.log('result3:', result3)

```

EDIT: added copying the all existing scope variables before passing them to `result += expr.evaluate(...)`

**Reply by josdejong** [reply]
On a side note: your solution using a replaceAll is tricky:

```js
mathjs.evaluate(expr.replaceAll(n_Symbol, i).replaceAll(k_Symbol, k_Value))
```
It replaces all characters, so if the expression is for example `sigmaSum("n=1", "k=5", "sin(n/pi) + k")`, it will replace the "n" in "sin" with the value `5`, making it `si5(5/pi) + k`.

Instead, you can pass the variables via a scope:

```js
mathjs.evaluate(expr, { 
  [n_Symbol]: i, 
  [k_Symbol]: k_Value
})
```

**Reply by NullDev** [reply]
This is amazing! Works flawlessly, even in combination with my other custom functions. Thank you so much! 
And yea, the `replaceAll` was a huge oversight 👀 
This is such an elegant solution. Probably should've read the docs carefully instead of overcomplicating stuff. I can't believe how well this works 

Thanks again!

**Reply by josdejong** [reply]
🎉 🚀  you're welcome. 

I use `rawArgs` for example to enable `plot(sin(x), x)` in [Math Notepad](https://mathnotepad.com/).

**Reply by NullDev** [reply]
I see! That's amazing :D 
I'm currently using it for an inline-math discord bot. It's a ton of hacky parser stuff but thanks to your solution, this is now a valid and computable expression: 

`f(x) = ∑(n=1, k=x-1, φ(n²)); f(10)`

the next step is to make 

`f(x) = ∑(n=1, k=x-1, ∑(a=n, b=x, φ(n² + a))); f(10)`

work. But ill try that on my own first :)

**Reply by josdejong** [reply]
😎

**Reply by NullDev** [reply]
Oh well. Haven't managed to do that yet. Do you maybe have a pointer for that or should I start a new discussion? 
Normal nesting works, e.g.: 

`∑(n=1, k=5, ∑(a=1, b=5, 2 + a))` -> 125

but as soon as I try to use another variable from "above" like:

`∑(n=1, k=5, ∑(a=1, b=5, n + a))` or `f(x) = ∑(n=1, k=5, ∑(a=1, b=5, x + a)); f(10)`

I get "Undefined symbol: n" and "Undefined symbol: x" respectively
What I tried:

```js
for (let i = n; i <= k; i++){
    const newScope = Object.assign({}, scope, { [nName]: i, [kName]: k });
    result += Number(expr.evaluate(newScope));
}
```

and 

```js
for (let i = n; i <= k; i++){
    const newScope = { ...scope };
    newScope[nName] = i;
    newScope[kName] = k;

    result += Number(exprNode.evaluate(newScope));
}
```

but neither worked

**Reply by josdejong** [reply]
Good point. The issue here is that `scope` is not an `Object` but a `Map`-like instance, so the spread operator doesn't work.

you can do something like this:

```js
for (let i = n; i <= k; i++) {
  const newScope = new Map()
  scope.forEach((value, key) => newScope.set(key, value))
  newScope.set(nName, i)
  newScope.set(kName, k)
  result += expr.evaluate(newScope)
}
```
I'll update my example with this fix.

Now, a simpler `const newScope = new Map(scope)` should also work (instead of the `forEach`), but for some reason this gives an issue, that is a bug in the `PartitionedMap` isntance that mathjs creates and uses. I'll have a look into that.

**Reply by NullDev** [reply]
Works like a charm. Thank you so much for your help (again 😅) and your amazing work on the library!

**Reply by josdejong** [reply]
Thanks, you're welcome 😊

**Reply by josdejong** [reply]
The issue with `new Map(scope)` is fixed via https://github.com/josdejong/mathjs/commit/a7f84abe51a72b6037c7fdf72c9295c9b18830a8 (not yet published)

**Reply by josdejong** [reply]
The fix for `new Map(scope)` is now published in `v12.4.0`

**Reply by NullDev** [reply]
Awesome, thanks!
